### PR TITLE
ui: Deployment Resources

### LINKED ISSUE #2263: ui: Deployment Resources
**Is your feature request related to a problem? Please describe.**
Now that Waypoint provides [detailed information on resources](https://github.com/hashicorp/waypoint/blob/5cbd52ff7a78ebe0594f09127c7f08788e2f51de/internal/server/proto/server.proto#L3235-L3287), I’d like to see that information in the web UI.

**Describe the solution you'd like**
On the deployment detail page, I’d like to see a list of the resources that comprise the deployment:

image

**Describe alternatives you've considered**
Resources could appear in various other places, but this seems like a good place to start.

**Explain any additional use-cases**
I want to see resources so that I can better understand how my app runs on the given platform, and how to debug it when things go wrong.

**Any additional context**
If you have access to it, see [WP-075](https://docs.google.com/document/d/1L1Scl1pLdakLkdx9AXru2CBUNh2tuB35YMnt_wu4qDQ/edit) for more background of the resource-awareness functionality.

### PR DESCRIPTION
**Pull request by jgwhite**
## Why the change?

Closes #2263 

This PR introduces resources to the web UI. Resources are elements of the underlying platform (i.e. Kubernetes pods, deployments) that are wholly or partially managed by Waypoint.

## What’s the plan?

- [x] Changelog entry
- [x] ~Start using `UI_ListDeployments` so we have `latestStatusReport` preloaded~
- [x] Restructure the app to allow for `deployment.resources...` sub-routes
- [x] Add deployment resource list
- [x] Add deployment resource detail
- [x] Add all the missing Mirage bits and bobs
- [x] Tidy up styles
- [x] Write some acceptance tests
- [x] Add metadata from designs
- [x] Make the state JSON collapsible (and collapsed by default?)
- [x] Include more-realistic state JSON in the Mirage setup
- [x] Empty states
- [x] Error states
- [x] Add more documentation for `includes` helper (particularly why we replace the ember-composable-helpers one)

### Light Mode

https://user-images.githubusercontent.com/34030/134903475-864eb0a4-f1d3-4010-88d1-dfa1436c9164.mp4

### Dark Mode

https://user-images.githubusercontent.com/34030/134903541-602d7f81-bdbd-4704-8242-679ac1ec10fd.mp4

### Screen Reader

https://user-images.githubusercontent.com/34030/134903599-e378aef7-66d7-480a-b1cc-ab8494812447.mp4

### Empty State

image

### Error State

image

### Using Mirage

1. Check out the branch:
   ```sh
   git checkout ui/resources
   ```
2. Boot the dev server
   ```sh
   cd ui && ember serve
   ```
3. [Visit the app](http://localhost:4200)
4. Visit some deployments
5. Verify that the new resources UI work well
6. Verify no existing functionality is obviously broken

### For realsies

1. Check out the branch:
   ```sh
   git checkout ui/resources
   ```
2. Build the ember app
   ```sh
   (cd ui && make)
   ```
3. Bundle the static assets
   ```sh
   make static-assets
   ```
4. Build the dev server
   ```sh
   make docker/server
   ```
5. Build the CLI
   ```sh
   make bin
   ```
6. Install waypoint on your platform of choice, i.e.
   ```sh
   ./waypoint install -platform=kubernetes -accept-tos -k8s-server-image=waypoint:dev -k8s-odr-image=waypoint-odr:dev
   ```
7. Open the UI
   ```
   ./waypoint ui -authenticate
   ```
8. Try out an example app of your choice (i.e. [kubernetes/nodejs](https://github.com/hashicorp/waypoint-examples/tree/main/kubernetes/nodejs))
9. Verify that the new resources UI work well
10. Verify no existing functionality is obviously broken

### CODE REVIEW

**Comment by almonk** [comment]
Generally this is looking great.

Some thoughts;

**Hide resources block when resources == 0**
When no resources are present, should we hide the resources block entirely? Presumably this is the case for any provider that isn't Kubernetes or ECS right now. Showing an empty state implies that the user needs to do something when in fact they may not.

**Semantics of "age"**
Right now our semantics around age are a bit odd: "Age: half a minute ago". This is a bit weird to read. We should either remove "ago" or change `Age`→`Created`.

**Comment by jgwhite** [comment]
> **Hide resources block when resources == 0**
> When no resources are present, should we hide the resources block entirely? Presumably this is the case for any provider that isn't Kubernetes or ECS right now. Showing an empty state implies that the user needs to do something when in fact they may not.
> 
> **Semantics of "age"**
> Right now our semantics around age are a bit odd: "Age: half a minute ago". This is a bit weird to read. We should either remove "ago" or change `Age`→`Created`.

@almonk both of these sound good to me. Will update.

**Comment by almonk** [comment]
@jgwhite nice – can we co-ordinate `Age`→`Created` for the CLI too so that the UIs are consistent? cc @briancain

File: ui/app/helpers/includes.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export default helper(([haystack, needle]: [unknown, unknown]): boolean => {
+  if (isHaystack(haystack)) {
+    return haystack.includes(needle);
+  } else {
+    return false;
+  }
+});
+
+function isHaystack(haystack?: unknown): haystack is Haystack {
+  return !!haystack && typeof (haystack as Haystack).includes === 'function';
```
Comment (gregone) [inline ui/app/helpers/includes.ts:16]: reading this code made me laugh 😄 🪡 . Just one suggestion though: if we're going to keep the helper as is, maybe add a comment to explain when to use it?

**Review by gregone — APPROVED** [review]
Looks great. 👍 
Just on minor comment about clarity for the `includes` helper.

File: ui/app/helpers/includes.ts (hunk 1 above)
Comment (jgwhite) [reply ui/app/helpers/includes.ts:16]: Good point! ember-composable-helpers also has [an `includes` helper](https://github.com/DockYard/ember-composable-helpers#includes) but it *only* works on arrays. Will add some more docs and usage guidance 👍

**Pull request merged**